badrcptto.c

```diff
@@ -0,0 +1,67 @@
+#include <unistd.h>
+#include "byte.h"
+#include "constmap.h"
+#include "control.h"
+#include "env.h"
+#include "fmt.h"
+#include "str.h"
+#include "stralloc.h"
+#include "strerr.h"
+
+extern void die_control();
+extern void die_nomem();
+
+static void _badrcptto_log_rejection(char *recipient)
+{
+  char smtpdpid[32];
+  char *remoteip = env_get("TCPREMOTEIP");
+  if (!remoteip) remoteip = "unknown";
+  str_copy(smtpdpid + fmt_ulong(smtpdpid,getppid())," ");
+  strerr_warn5("rcptcheck: badrcptto ",smtpdpid,remoteip," ",recipient,0);
+}
+
+static int _badrcptto_reject_exact_address(struct constmap map, stralloc address)
+{
+  return (1 && constmap(&map,address.s,address.len - 1));
+}
+
+static int _badrcptto_reject_whole_domain(struct constmap map, stralloc address)
+{
+  /* why not just comment out the domain in control/rcpthosts? */
+  int j = byte_rchr(address.s,address.len,'@');
+  return ((j < address.len) && (constmap(&map,address.s + j,address.len - j - 1)));
+}
+
+static int _badrcptto_reject_string(char *string)
```

Comment:
A `char *` string passed without a length looks suspicious: would `char *string` be written to without a length? Of course not, it is only read.

Making it `char const *` saves the time it takes to check it, and make it more obvious that the function returns 0 or 1 depending on `string`.